### PR TITLE
IDEMPIERE-5224 - Locator Field type Caching Improvement

### PR DESCRIPTION
**Pull request by PeterTakacs300**
Ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5224

Implemented changes made by @igorpojzl

### CODE REVIEW

File: org.adempiere.base/src/org/compiere/model/MLocatorLookup.java [hunk 1]
```diff
@@ -483,12 +486,13 @@ public void run()
...
-		ArrayList<Object> list = new ArrayList<Object>(collection.size());
-		Iterator<MLocator> it = collection.iterator();
+		Collection<KeyNamePair> collection = getData();
+		ArrayList<Object> list = new ArrayList<Object>(collection);
+		Iterator<KeyNamePair> it = collection.iterator();
 		while (it.hasNext())
 		{
-			MLocator loc = it.next();
+			KeyNamePair locPair = it.next();
+			MLocator loc = MLocator.get(locPair.getKey());
```
Comment (PeterTakacs300) [inline org.adempiere.base/src/org/compiere/model/MLocatorLookup.java:495]: To improve caching, this needs to be removed.

File: org.adempiere.base/src/org/compiere/model/MLocatorLookup.java [hunk 2]
```diff
@@ -483,12 +486,13 @@ public void run()
...
 			if (isValid(loc))				//	only valid warehouses
 				list.add(loc);
```
Comment (PeterTakacs300) [inline org.adempiere.base/src/org/compiere/model/MLocatorLookup.java:497]: To improve caching, this needs to be removed.

**Comment by hengsin** [comment]
hi @PeterTakacs300 , to facilitate review, please give a brief description on the solution to the issue mention in the ticket, thanks.

**Comment by norbertbede** [comment]
This was originally developer by @igorpojzl. Please provide required informations

**Comment by igorpojzl** [comment]
Goul in This PR is to remove all MLocator.get calls in MLocatorLookup.

MLocatorLookup is Storing MLocators(cache) to its collection which can cause to overflow MLocator Cache (between multiple clients). Which cause high volume of calls to Database. 
We prepared to change LocatorLookup to store KeyNamePairs.

There is still question and issue with Java method 
**public ArrayList<Object> getData (boolean mandatory, boolean onlyValidated, boolean onlyActive, boolean temporary, boolean shortlist) // IDEMPIERE 90**

which is loading MLocators from cache to validate each locator, and negating change in this PR.

**Comment by hengsin** [comment]
Hi @igorpojzl , if the goals is to remove MLocator.get() db call, perhaps we should change the Loader to select * from MLocator and we can use the ResultSet constructor of MLocator instead ? That would works with the isvalid call too.

**Comment by igorpojzl** [comment]
hi @hengsin we also removed logic for only stored product to allow set any locator for Product.

**Comment by hengsin** [comment]
hi @igorpojzl , it seems the isValid method is make dummy now. If that's the case, we should mark that as deprecated and remove all call of that method from cire.

**Comment by hengsin** [comment]
hi @CarlosRuiz-globalqss , this looks fine to me but I'm not sure whether the isValid change will break anything, wdyt ?

**Comment by hengsin** [comment]
hi @CarlosRuiz-globalqss , this looks fine to me but I'm not sure whether the isValid change will break anything, wdyt ?

**Comment by CarlosRuiz-globalqss** [comment]
Hi team,

Is not easy to track the impact of the changes.

From what I see, it seems the "only_Product" feature of locator editor is destroyed.

I followed the test case described by @hengsin at [IDEMPIERE-4382](https://idempiere.atlassian.net/browse/IDEMPIERE-4382) and it happens exactly what is described there, so this is creating a regression about that fix, and I think is because the locator doesn't respect now the "only product" stuff.


I have one doubt, according to the mattermost conversation, this is caused by REST loading the lookup, this is on the window rest?  Asking because I have not used those, just the model part.


Also, can you please describe better what is the issue?  Is MLookup loading the full list of 6.000 locators in memory?  If that's the case, how the changes make that better, how is not loading the same number of MLocator records now?


Sorry, I'm a bit lost jumping so late here.

Regards,

Carlos Ruiz

**Comment by norbertbede** [comment]
Facts: 17 000 locators, 10 000 cache size, approx 6000/most active 2 clients. this produced approx 1mill calls.
we analyze cache in idempiere monitor and first 1 product call loaded 6000 locators to cache, next used it. 2nd tenant made eg. material receipt in webui loaded next 6000 so first 2000 was destroyed in cache.

we are sure this happened in REST model calls, and webui also. 
(i made 2 upgrades ferst rest, significant reduction then 2nd day webui, then goes down for )

After @igorpojzl  changes we save daily approx 900 000 db calls.

So the code is good for me :) Cant comment it.

1 more issue: Locator fieldtype lists only locators with storage levels, we changed this also. because the locator in product s for set default, this is considered in code in many places.

**Comment by hengsin** [comment]
hi @norbertbede , @igorpojzl  & @CarlosRuiz-globalqss ,
I've implemented an alternative approach at pull/1242 with the following difference:
- use MLocator ResultSet constructor and is backward compatible
- implemented lazy loading of MLocatorLookup. With lazy loading, the locator editor wouldn't pull all locator records from db until the user open the locator dialog.

I've not done extensive testing with the lazy loading part so that might have issue(s) that I've not discover yet.

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]
Comments on ticket:
https://idempiere.atlassian.net/browse/IDEMPIERE-5224?focusedCommentId=48122

This change is reopening an old fixed bug.

**Comment by CarlosRuiz-globalqss** [comment]
Replaced with https://github.com/idempiere/idempiere/pull/1242

**Pull request closed without merging**